torch/fx/experimental/symbolic_shapes.py

```diff
@@ -467,7 +468,13 @@ def _print_Symbol(self, expr) -> str:
 
 class ShapeEnv(object):
     def __init__(self):
-        self.guards = []
+        # TODO(voz): For real prod cases downstream of dynamo, we should always have a tracing context
+        # available from .get()
```

Comment:
If we're consolidating in this way, we really shouldn't have ShapeEnv generating its own list. Instead this should just be a hard assert error. But it also feels like there's a layering problem again here: guards is not the "only" extractable state here; for example, it doesn't make sense to look at the guards without knowing other things about the variables; e.g., `val-to_var`, or the `tls` which lets you suppress shape guards. So it really seems more like ShapeEnv should be on the context itself, and then its guards field is the canonical source of shape guards.

Reply:
I agree - this was my plan for the checkpointing PR - to move ShapeEnv onto TracingContext, after the FakeTensorMode move. 

Open to suggestions about how to better organize it. 

This is a bit of an intermediary state to get things going. It need not be like this if you hate it though. We could also Make ShapeEnv take a TracingContext in the ctor as a stopgap before correcting all layering and moving it there, as thats a bigger thing on its own.

Reply:
What I suggest, is drop this part of the patch, so that this change is ONLY introducing ShapeGuard struct. I will happily approve that. Then, do the ShapeEnv move all in one go, without this intermediate

Reply:
Sgtm. Do you care if I move shape env off of fake tensor mode in the same go? I feel like we should not have both a tracing context and a mode be a way to get the current shape env. 

It's kind of a big change, and I need to think through the mode construction story a bit more, but it feels correct. Thoughts?

Reply:
Wait, isn't fake tensor mode going on the tracing context? So you don't have to move ShapeEnv off it at all...

FakeTensorMode legitimately needs access to the ShapeEnv when you are doing fake tensor stuff with dynamic shapes, but a tracing context is not guaranteed to exist (unless, I guess, if you're planning to force one to exist). So it seems better to maintain TracingMode -> FakeTensorMode -> ShapeEnv.

Reply:
Are you okay with the fake tensor mode registering itself with the tracing context if one is around?

Reply:
That doesn't make much sense to me. In Dynamo today, there is no registering itself; Dynamo allocates a fake mode, and then we consistently use the fake mode on the tx whenever we need a fake mode. Morally, the tracing context is precisely that, just in a place that more than dynamo can use it.